### PR TITLE
Allow to rename uploaded files using common patterns

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

If you read the [SymfonyDocs about uploading files](https://symfony.com/doc/current/controller/upload_file.html), you'll find code like this:

```php
$originalFilename = pathinfo($uploadedFile->getClientOriginalName(), PATHINFO_FILENAME);
$safeFilename = transliterator_transliterate('Any-Latin; Latin-ASCII; [^A-Za-z0-9_] remove; Lower()', $originalFilename);
$newFilename = $safeFilename.'-'.uniqid().'.'.$uploadedFile->guessExtension();

$uploadedFile->move('/path/to/uploads', $newFilename);
``` 

This code is verbose and repetitive, so I propose to add a little method to `UploadFile` class to simplify the previous code. If this PR is approved, the previous code will look like this:

```php
$uploadedFile->moveAndRename(
    '/path/to/uploads',
    '[slug]-[randomhash].[extension]'
);
```

The new method would be small, but it will allow lots of different solutions:

```php
$uploadedFile->moveAndRename('...', '[name]-[timestamp].[extension]');
$uploadedFile->moveAndRename('...', 'photo-[contenthash].[extension]');
$uploadedFile->moveAndRename('...', '[year]/[month]/file-[day]-[randomhash].[extension]');
$uploadedFile->moveAndRename('...', '[uuid].[extension]');
$uploadedFile->moveAndRename('...', 'contract-[randomhash]-[timestamp].[extension]');
```

## Next steps

If this proposal is approved:

* I'll finish it adding tests and docs.
* We need to think if the given patterns are enough to cover the most common cases.
* Should we allow defining the `[extension]` pattern .. or should we always add the guessed extension as a mandatory behavior?
* We need to think if adding a new method (`moveAndRename()`) is OK or if we could solve this differently with the existing methods.

Thanks!

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/File/UploadedFile.php [hunk 1]
```diff
@@ -236,6 +236,39 @@ public function move($directory, $name = null)
         throw new FileException($this->getErrorMessage());
     }
 
+    public function moveAndRename(string $targetDirectory, string $fileNamePattern)
+    {
+        $fileName = str_replace(
```
Comment (yceruto) [inline src/Symfony/Component/HttpFoundation/File/UploadedFile.php]: could we use `strtr` for a better key/value visualization?

**Comment by yceruto** [comment]
Another alternative might be to add it as the third argument of the `move` method:
```php
public function move($directory, $name = null, string $pattern = null)
```

File: src/Symfony/Component/HttpFoundation/File/UploadedFile.php [hunk 2]
```diff
@@ -236,6 +236,39 @@ public function move($directory, $name = null)
         throw new FileException($this->getErrorMessage());
     }
 
+    public function moveAndRename(string $targetDirectory, string $fileNamePattern)
```
Comment (yceruto) [inline src/Symfony/Component/HttpFoundation/File/UploadedFile.php]: we might also need to add to the phpdoc all the placeholders available for that argument.

**Comment by yceruto** [comment]
I'd like to add the `$pattern` argument to the current `move()` method for 5.0.

Something like this:
```php
public function move($directory, $name = null/*, string $pattern = null */)
```
and using one or the other,

or this one to process the `$name` argument as a name pattern or not:
```php
public function move($directory, $name = null/*, bool $pattern = false */)
```

**Comment by javiereguiluz** [comment]
I don't like mixing both "name" and "name pattern" in the same method ... because it's confusing: can I use both at the same time? The pattern is applied on the original name or on the passed name? etc.

We could do this too:

* Deprecate the "move and rename" behavior of `move()` and allow it to only "move" files.
* Introduce this "moveAndRename()" method to mimic the old "move()" behavior, but directly supporting renaming based on patterns.

File: src/Symfony/Component/HttpFoundation/File/UploadedFile.php [hunk 3]
```diff
@@ -236,6 +236,24 @@ public function move($directory, $name = null)
...
+    {
+        $fileName = strtr($fileNamePattern, [
+            '[contenthash]' => sha1_file($this->getRealPath()),,
+            '[day]' => date('d'),
+            '[extension]' => $this->getExtension(),
+            '[month]' => date('m'),
+            '[name]' => $this->getBasename(),
+            '[randomhash]' => bin2hex(random_bytes(20)),
+            '[slug]' => transliterator_transliterate('Any-Latin; Latin-ASCII; [^A-Za-z0-9_] remove; Lower()', $this->getBasename()),
```
Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/File/UploadedFile.php]: this requires the intl-extension, which is not stubbed by symfony/intl (nor is a composer dev-dependency for this component)

File: src/Symfony/Component/HttpFoundation/File/UploadedFile.php [hunk 4]
```diff
@@ -294,4 +312,15 @@ public function getErrorMessage()
 
         return sprintf($message, $this->getClientOriginalName(), $maxFilesize);
     }
+
+    private function generateUuid4()
```
Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/File/UploadedFile.php]: should be static IMHO

File: src/Symfony/Component/HttpFoundation/File/UploadedFile.php [hunk 5]
```diff
@@ -236,6 +236,24 @@ public function move($directory, $name = null)
         throw new FileException($this->getErrorMessage());
     }
 
+    public function moveAndRename(string $targetDirectory, string $fileNamePattern)
+    {
+        $fileName = strtr($fileNamePattern, [
+            '[contenthash]' => sha1_file($this->getRealPath()),,
```
Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/File/UploadedFile.php]: i think we should prefer a reserved character, e.g. `%key%`

**Review by ro0NL — COMMENTED** [review]
im a bit worried about the extendability of such a feature, and thus its value:/

what if i want a longer/shorter hash, use `j` instead of `d` for a date format, etc.

**Comment by javiereguiluz** [comment]
@ro0NL you are right, this is not infinitely flexible. However:

* Previous situation: 100% of developers uploading files must add some code themselves.
* New situation: 85% of developers can use this feature to meet their needs. The other 15% continue in the previous situation.

Also, this could be an interesting simplification for other projects. For example, VichUploaderBundle defines hundreds of lines of code and multiple classes to create "namers" (https://github.com/dustin10/VichUploaderBundle/blob/master/Resources/doc/namers.md) which are arguably less flexible than this proposed method.

Finally, I proposed to use `[name]`, `[contenthash]`, etc. instead of `%name%`, `%content_hash%` because the former is natural to lots of developers. It's for example what we use in Encore because of Webpack (see https://webpack.js.org/guides/caching/)

**Comment by stof** [comment]
Should this be a method of UploadedFile, or a separate utility service (which could allow extension points in the future to register additional placeholders for instance)

File: src/Symfony/Component/HttpFoundation/File/UploadedFile.php (hunk 3 above)
Comment (stof) [reply src/Symfony/Component/HttpFoundation/File/UploadedFile.php]: and the way things are implemented, this is a dependency even if they don't use the `[slug]` placeholder.

File: src/Symfony/Component/HttpFoundation/File/UploadedFile.php [hunk 6]
```diff
@@ -294,4 +312,15 @@ public function getErrorMessage()
 
         return sprintf($message, $this->getClientOriginalName(), $maxFilesize);
     }
+
+    private function generateUuid4()
+    {
+        return sprintf('%04x%04x-%04x-%04x-%04x-%04x%04x%04x',
+            mt_rand(0, 0xffff), mt_rand(0, 0xffff),
```
Comment (stof) [inline src/Symfony/Component/HttpFoundation/File/UploadedFile.php]: shouldn't it use `random_int` instead of `mt_rand` ?

**Comment by javiereguiluz** [comment]
@stof in my opinion, this shouldn't be extensible in the way you are proposing (similar to EnvVar processors). Creating an extension to add a placeholder would take much more effort than just writing the logic to do your custom naming.

**Comment by stof** [comment]
well, in your own project, indeed. But maybe some bundles could share some of them.

**Comment by OskarStark** [comment]
Focusing on the value I really like this kind of pattern for filenames, but I agree with @stof that having extension points would be nice.

Anyway for the start some are enough to see if it gets adopted.

👍🏻 for not having it in UploadedFile

I think that the con of @ro0NL with the longer/shorter hash is not big enough to not implement such thing.

Thanks for the proposal Javier 👏

File: src/Symfony/Component/HttpFoundation/File/UploadedFile.php (hunk 5 above)
Comment (OskarStark) [reply src/Symfony/Component/HttpFoundation/File/UploadedFile.php]: One comma is enough here

**Comment by nicolas-grekas** [comment]
:-1: on my side sorry. Note that the example above should be rewritten to use the new String component now, this would be much cleaner.
If we really want this logic, it could be moved to a separate static method/helper.
`$uploadedFile->move($dir, SomeHelper::format($uploadedFile, $pattern));`

**Comment by javiereguiluz** [comment]
We implemented this feature a while ago in EasyAdmin and I opened this in case there was some interest in adding this feature to Symfony too. But, most people doesn't seem interested ... and others want to change too many things ... so I agree that the best thing is closing it. Thanks!

**Pull request closed without merging**